cmd/snap-confine/mount-support.c

```diff
@@ -82,21 +82,42 @@ static void setup_private_mount(const char *snap_name)
 	sc_must_snprintf(base_dir, sizeof(base_dir), "/tmp/snap.%s", snap_name);
 	sc_must_snprintf(tmp_dir, sizeof(tmp_dir), "%s/tmp", base_dir);
 
+	// NOTE: Because snapd executes with group ownership of the calling
+	// user, temporarily change back to root. This workaround code can be
+	// removed once central handling of uid/gid is merged into
+	// snap-confine.
+	gid_t old_gid = getgid();
+	if (setgid(0) < 0) {
+		die("cannot set gid to 0");
+	}
 	// Create /tmp/snap.$SNAP_NAME/ 0700 root.root. Ignore EEXIST since we want
 	// to reuse and we will open with O_NOFOLLOW, below.
 	if (mkdir(base_dir, 0700) < 0 && errno != EEXIST) {
 		die("cannot create base directory %s", base_dir);
 	}
+	// Restore original gid, please see the note above.
+	if (setgid(old_gid) < 0) {
+		die("cannot restore gid to %d", old_gid);
+	}
+
 	base_dir_fd = open(base_dir,
 			   O_RDONLY | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
 	if (base_dir_fd < 0) {
 		die("cannot open base directory %s", base_dir);
 	}
-	if (fchmod(base_dir_fd, 0700) < 0) {
-		die("cannot chmod base directory %s to 0700", base_dir);
+	struct stat file_info;
+	if (fstat(base_dir_fd, &file_info) < 0) {
+		die("cannot stat base directory %s", base_dir);
 	}
-	if (fchown(base_dir_fd, 0, 0) < 0) {
-		die("cannot chown base directory %s to root.root", base_dir);
+	debug("base dir %d %d %4o", file_info.st_uid, file_info.st_gid,
+	      file_info.st_mode & 07777);
+	if (file_info.st_uid != 0 || file_info.st_gid != 0) {
+		die("cannot proceed with non-root-owned base directory %s",
+		    base_dir);
+	}
+	if ((file_info.st_mode & 07777) != 0700) {
+		die("cannot proceed with non-0700-mode base directory %s",
+		    base_dir);
```

Comment:
I was thinking about this. Perhaps snapd should move the original entry aside and create a temporary directory and rename it back. We can verify that the inode is the same after the rename. My main motivation is to preserve the user experience of "users can know where to find files that snaps placed in /tmp".

Reply:
Well, the user experience would remain the same in the normal case; it's only under the shenanigans case that the fallback behavior is used, which should be fine, because we appropriately complained about it so an admin can start investigating.

If you move it aside and put in in place, there is a race there, so you'll be faced with the same problem. Failing is safe but then there is a DoS (and thus a CVE because this is about an unprivileged user not just DoSing herself (not a CVE) but DoSing use of the snap by all users (a privileged action, thus CVE)). When using a fallback, the whole thing is safe, predictable and no DoS to prevent mission critical applications, etc.

Reply:
Thank you for explaining the importance of non-privileged user being able to DOS other users on the system. Let me think about this some more. I'm not super fond of the fallback case because it is a special case that will not be used often and the UX will not be great either (the user may easily not notice the complain and we don't have a way to use the warnings system from C.